### PR TITLE
[risk=low][RW-1977][RW-3313][RW-3314] Resource Card refactoring 

### PR DESCRIPTION
**Pull request by jmthibault79**
Added new-style (based on ResourceCardTemplate) components for the remaining two resource types (Concept Set and Cohort Review) and deleted the old ResourceCard.

Previous PRs in this series are #2327 #2493 #3053

Added resource-related utilities:
- DataSetReferenceModal
- ResourceActionsMenu
- getResourceCard
- resources.ts

Description:

---
**PR checklist**

- [x] This PR meets the Acceptance Criteria in the JIRA story
- [x] The JIRA story has been moved to Dev Review
- [x] This PR includes appropriate unit tests
- [x] I have run and tested this change locally
- [x] I have run the E2E tests on ths change against my local UI + API server with `yarn test-local`
- [x] If this includes a UI change, I have taken screen recordings or screenshots of the new behavior and notified the PO and UX designer
- [x] If this includes an API change, I have updated the appropriate Swagger definitions and notified API consumers
- [x] If this includes a new feature flag, I have created and linked new JIRA tickets to (a) turn on the feature flag and (b) remove it later

### CODE REVIEW

File: ui/src/app/components/data-set-reference-modal.tsx [hunk 1]
```diff
@@ -0,0 +1,75 @@
+import * as fp from 'lodash/fp';
```
Comment (jmthibault79) [inline ui/src/app/components/data-set-reference-modal.tsx:1]: Extracted from ReferenceCard, with some updates

File: ui/src/app/components/resource-card-template.tsx [hunk 2]
```diff
@@ -54,28 +64,60 @@ const styles = reactStyles({
...
-export interface Action {
+interface Action {
   icon: string;
   displayName: string;
   onClick: () => void;
   disabled: boolean;
   hoverText?: string;
 }
 
+const ResourceActionsMenu = (props: {actions: Action[]}) => {
```
Comment (jmthibault79) [inline ui/src/app/components/resource-card-template.tsx:75]: moved up from below in the file, and made into a component.  I'll reuse this for RW-5562.

File: ui/src/app/components/resource-card-template.tsx [hunk 3]
```diff
@@ -54,28 +64,60 @@ const styles = reactStyles({
...
+      }
+  >
+    <Clickable data-test-id='resource-menu'>
+      <SnowmanIcon disabled={false}/>
+    </Clickable>
+  </PopupTrigger>;
+};
+
+
+function canWrite(resource: WorkspaceResource): boolean {
```
Comment (jmthibault79) [inline ui/src/app/components/resource-card-template.tsx:104]: common code from the individual resource-type cards

File: ui/src/app/components/resource-card-template.tsx [hunk 4]
```diff
@@ -54,28 +64,60 @@ const styles = reactStyles({
...
+}
+
+function canDelete(resource: WorkspaceResource): boolean {
+  return resource.permission === 'OWNER';
+}
+
 interface Props {
   actions: Action[];
   disabled: boolean;
-  resourceUrl: string;
```
Comment (jmthibault79) [inline ui/src/app/components/resource-card-template.tsx:68]: can simplify the interface b/c all of these are determinable from the resource

File: ui/src/app/components/resource-card-template.tsx [hunk 5]
```diff
@@ -54,28 +64,60 @@ const styles = reactStyles({
...
+  resource: WorkspaceResource;
   onNavigate: () => void;
-  displayName: string;
-  description: string;
-  displayDate: string;
-  footerText: string;
-  footerColor: string;
 }
 
 // This will be renamed to ResourceCard once the old code is removed
```
Comment (jmthibault79) [inline ui/src/app/components/resource-card-template.tsx:119]: TODO ASAP.  Not in this PR because I think that will confuse things.

File: ui/src/app/components/resources.tsx [hunk 6]
```diff
@@ -1,71 +0,0 @@
...
-export interface ResourceCardMenuProps {
-  resourceType: ResourceType;
-  onRenameResource?: Function;
-  onCopyConceptSet?: Function;
-  canDelete: boolean;
-  onDeleteResource?: Function;
-  canEdit: boolean;
-}
-
-export class ResourceCardMenu extends React.Component<ResourceCardMenuProps> {
```
Comment (jmthibault79) [inline ui/src/app/components/resources.tsx:18]: Replaced by `ResourceActionsMenu` and the resource types' `get actions()`

File: ui/src/app/pages/analysis/notebook-list.tsx [hunk 7]
```diff
@@ -26,6 +25,19 @@ const styles = {
   }
 };
 
+function convertToResource(inputResource: FileDetail, workspace: WorkspaceData): WorkspaceResource {
```
Comment (jmthibault79) [inline ui/src/app/pages/analysis/notebook-list.tsx:28]: only used by Notebooks and tests, so moved it here

File: ui/src/app/pages/analysis/notebook-resource-card.tsx [hunk 8]
```diff
@@ -42,32 +45,8 @@ export const NotebookResourceCard = fp.flow(
     };
   }
 
-  get resourceType(): ResourceType {
```
Comment (jmthibault79) [inline ui/src/app/pages/analysis/notebook-resource-card.tsx:45]: Lots of common code pushed down to the template.

Many of the actions are very similar across resources, so there's probably an opportunity to reduce the code further.  I'm not planning to do that here.

File: ui/src/app/pages/data/cohort-review/cohort-review-resource-card.tsx [hunk 9]
```diff
@@ -0,0 +1,107 @@
...
+  resource: WorkspaceResource;
+  existingNameList: string[];
+  onUpdate: Function;
+}
+
+interface State {
+  showRenameModal: boolean;
+}
+
+export const CohortReviewResourceCard = fp.flow(
```
Comment (jmthibault79) [inline ui/src/app/pages/data/cohort-review/cohort-review-resource-card.tsx:23]: https://precisionmedicineinitiative.atlassian.net/browse/RW-3313

File: ui/src/app/pages/data/cohort/cohort-resource-card.tsx [hunk 10]
```diff
@@ -20,6 +20,7 @@ interface Props extends WithConfirmDeleteModalProps, WithErrorModalProps, WithSp
 
 interface State {
   showRenameModal: boolean;
+  referencingDataSets: Array<DataSet>;
```
Comment (jmthibault79) [inline ui/src/app/pages/data/cohort/cohort-resource-card.tsx:23]: Restores the functionality from https://precisionmedicineinitiative.atlassian.net/browse/RW-2945 which was accidentally removed in a previous refactor

File: ui/src/app/pages/data/cohort/cohort-resource-card.tsx [hunk 11]
```diff
@@ -31,89 +32,87 @@ export const CohortResourceCard = fp.flow(
...
 
-  delete() {
-    return cohortsApi().deleteCohort(
-      this.props.resource.workspaceNamespace,
-      this.props.resource.workspaceFirecloudName,
-      this.props.resource.cohort.id)
-      .then(() => {
-        this.props.onUpdate();
-      });
+  async setOrClearDataSetList() {
```
Comment (jmthibault79) [inline ui/src/app/pages/data/cohort/cohort-resource-card.tsx:88]: The logic in this extracted and renamed code was hard to follow.  I may try to simplify it.

File: ui/src/app/pages/data/concept/concept-set-resource-card.tsx [hunk 12]
```diff
@@ -0,0 +1,166 @@
...
+  onUpdate: Function;
+}
+
+interface State {
+  showRenameModal: boolean;
+  copyingConceptSet: boolean;
+  referencingDataSets: Array<DataSet>;
+}
+
+export const ConceptSetResourceCard = fp.flow(
```
Comment (jmthibault79) [inline ui/src/app/pages/data/concept/concept-set-resource-card.tsx:27]: https://precisionmedicineinitiative.atlassian.net/browse/RW-3314

File: ui/src/app/utils/resourceActions.ts [hunk 13]
```diff
@@ -1,79 +0,0 @@
...
-  ConceptSet,
-  DataSet,
-  FileDetail,
-  ResourceType,
-  WorkspaceAccessLevel,
-  WorkspaceResource
-} from 'generated/fetch';
-import {WorkspaceData} from './workspace-data';
-
-export function toDisplay(resourceType: ResourceType): string {
```
Comment (jmthibault79) [inline ui/src/app/utils/resourceActions.ts:13]: moved to resources.ts

File: ui/src/app/utils/resourceActions.ts [hunk 14]
```diff
@@ -1,79 +0,0 @@
...
-  }
-}
-
-export interface ConvertToResourcesArgs {
-  list:  FileDetail[] | Cohort[] | CohortReview[] | ConceptSet[] | DataSet[];
-  resourceType: ResourceType;
-  workspace: WorkspaceData;
-}
-
-export function convertToResources(args: ConvertToResourcesArgs): WorkspaceResource[] {
```
Comment (jmthibault79) [inline ui/src/app/utils/resourceActions.ts:40]: moved to testing/stubs/resources-stub

File: ui/src/app/components/data-set-reference-modal.tsx [hunk 15]
```diff
@@ -0,0 +1,77 @@
...
+      await deleteResource();
+    } catch (ex) {
+      console.log(ex);
+    }
+  }
+
+  render() {
+    const {referencedResource, dataSets, onCancel} = this.props;
+
+    const resourceName = fp.startCase(getDisplayName(referencedResource));
```
Comment (jmthibault79) [inline ui/src/app/components/data-set-reference-modal.tsx:51]: New styling components/layout.

Old, from reference-card.tsx
```
<div style={{paddingBottom: '1rem'}}>	
            The {toDisplay(this.resourceType)} <b>{fp.startCase(this.displayName)}&nbsp;</b>	
            is referenced by the following datasets:	
            <b>	
              &nbsp;	
              {fp.join(', ' ,	
                this.state.dataSetByResourceIdList.map((data) => data.name))}	
            </b>.	
            Deleting the {toDisplay(this.resourceType)}	
            <b>{fp.startCase(this.displayName)} </b> will make these datasets unavailable for use.	
            Are you sure you want to delete <b>{fp.startCase(this.displayName)}</b> ?	
</div>
```

File: e2e/tests/notebook/notebook-owner-actions.spec.ts [hunk 16]
```diff
@@ -73,7 +73,7 @@ describe('Workspace owner Jupyter notebook action tests', () => {
 
     const newName = makeRandomName('test-notebook');
     const modalTextContents = await workspaceAnalysisPage.renameResource(notebookName, newName, ResourceCard.Notebook);
-    expect(modalTextContents).toContain(`Enter new name for ${notebookName}.ipynb`);
+    expect(modalTextContents).toContain(`Enter new name for ${notebookName}`);
```
Comment (jmthibault79) [inline e2e/tests/notebook/notebook-owner-actions.spec.ts:76]: Notebook names in resource actions are now uniform - no extension

File: ui/src/app/components/get-resource-card.tsx [hunk 17]
```diff
@@ -0,0 +1,33 @@
...
+import {isCohort, isCohortReview, isConceptSet, isDataSet, isNotebook} from 'app/utils/resources';
+import {BillingStatus, WorkspaceResource} from 'generated/fetch';
+
+interface GetResourceCardProps {
+  resource: WorkspaceResource;
+  existingNameList: string[];
+  onUpdate: () => Promise<void>;
+}
+
+function getResourceCard(props: GetResourceCardProps) {
```
Comment (ericsong) [inline ui/src/app/components/get-resource-card.tsx:18]: nit: I think "render" better captures what is happening here versus "get"

File: ui/src/app/utils/resources.tsx [hunk 18]
```diff
@@ -0,0 +1,95 @@
...
+  return fp.cond([
+      [isCohort, () => ResourceType.COHORT],
+      [isCohortReview, () => ResourceType.COHORTREVIEW],
+      [isConceptSet, () => ResourceType.CONCEPTSET],
+      [isDataSet, () => ResourceType.DATASET],
+      [isNotebook, () => ResourceType.NOTEBOOK],
+  ])(resource);
+}
+
+export {
```
Comment (ericsong) [inline ui/src/app/utils/resources.tsx:82]: can we also move get-resource-card in here? Since it seems like the place to place related functions.

File: ui/src/testing/stubs/resources-stub.ts [hunk 19]
```diff
@@ -0,0 +1,53 @@
...
+    ResourceType,
+    WorkspaceAccessLevel,
+    WorkspaceResource,
+} from 'generated/fetch';
+import {CdrVersionsStubVariables} from 'testing/stubs/cdr-versions-api-stub';
+import {WorkspaceStubVariables} from 'testing/stubs/workspace-service-stub';
+
+type InputResource = FileDetail | Cohort | CohortReview | ConceptSet | DataSet;
+
+export function convertToResource(
```
Comment (ericsong) [inline ui/src/testing/stubs/resources-stub.ts:18]: Can this be combined with the real implementation above? It looks like this function is a superset of the functionality that's in the notebook file.

**Review by ericsong — APPROVED** [review]
Looks great! awesome stuff

File: ui/src/app/components/data-set-reference-modal.tsx [hunk 20]
```diff
@@ -0,0 +1,75 @@
...
+  resource: {
+    fontWeight: 'bold',
+  },
+  dataSets: {
+    fontWeight: 'bold',
+  }
+});
+
+interface Props {
+  referencedResource: WorkspaceResource;
```
[the remark this answers is not in the document]
Comment (jmthibault79) [reply ui/src/app/components/data-set-reference-modal.tsx:21]: it's actually only Cohorts and Concept Sets, since these are what get referenced by Datasets.

File: ui/src/app/utils/resources.tsx (hunk 18 above)
Comment (jmthibault79) [reply ui/src/app/utils/resources.tsx:82]: That's what I had originally, but it makes a circular reference so I had to split it out.

File: ui/src/testing/stubs/resources-stub.ts (hunk 19 above)
Comment (jmthibault79) [reply ui/src/testing/stubs/resources-stub.ts]: yes - the original reason for separating these is no longer valid
